README.rst
=================
Cryoassess plugin
=================

This plugin provides a wrapper for `Cryoassess <https://github.com/cianfrocco-lab/Automatic-cryoEM-preprocessing>`_ software tools for automatic micrograph and 2D classes assessment.

.. image:: https://img.shields.io/pypi/v/scipion-em-cryoassess.svg
        :target: https://pypi.python.org/pypi/scipion-em-cryoassess
        :alt: PyPI release

.. image:: https://img.shields.io/pypi/l/scipion-em-cryoassess.svg
        :target: https://pypi.python.org/pypi/scipion-em-cryoassess
        :alt: License

.. image:: https://img.shields.io/pypi/pyversions/scipion-em-cryoassess.svg
        :target: https://pypi.python.org/pypi/scipion-em-cryoassess
        :alt: Supported Python versions

.. image:: https://img.shields.io/sonar/quality_gate/scipion-em_scipion-em-cryoassess?server=https%3A%2F%2Fsonarcloud.io
        :target: https://sonarcloud.io/dashboard?id=scipion-em_scipion-em-cryoassess
        :alt: SonarCloud quality gate

.. image:: https://img.shields.io/pypi/dm/scipion-em-cryoassess
        :target: https://pypi.python.org/pypi/scipion-em-cryoassess
        :alt: Downloads

Installation
-------------

You will need to use 3.0+ version of Scipion to be able to run these protocols. To install the plugin, you have two options:

a) Stable version

.. code-block::

   scipion installp -p scipion-em-cryoassess

b) Developer's version

   * download repository

    .. code-block::

        git clone -b devel https://github.com/scipion-em/scipion-em-cryoassess.git

   * install

    .. code-block::

       scipion installp -p /path/to/scipion-em-cryoassess --devel

Cryoassess software will be installed automatically with the plugin but you can also use an existing installation by providing *CRYOASSESS_ENV_ACTIVATION* (see below).
You also have to download training models separately (see below).

**Important:** you need to have conda (miniconda3 or anaconda3) pre-installed to use this program.

Configuration variables
-----------------------

*CONDA_ACTIVATION_CMD*: If undefined, it will rely on conda command being in the
PATH (not recommended), which can lead to execution problems mixing scipion
python with conda ones. One example of this could can be seen below but
depending on your conda version and shell you will need something different:
CONDA_ACTIVATION_CMD = eval "$(/extra/miniconda3/bin/conda shell.bash hook)"

*CRYOASSESS_ENV_ACTIVATION* (default = conda activate cryoassess-1.0.0):
Command to activate the cryoassess environment.

The deep-learning models can be downloaded from
`authors' website <https://cosmic-cryoem.org/software/cryo-assess/>`_ and the folder with models is set with:

*CRYOASSESS_MODELS* (default = software/em/cryoassess-models)

Verifying
---------

To check the installation, simply run the following Scipion test:

``scipion test cryoassess.tests.test_protocols_cryoassess.TestCryoassess``

Supported versions
------------------

1.0.0

Protocols
----------

* assess micrographs
* assess 2D classes

References
-----------

1. High-Throughput Cryo-EM Enabled by User-Free Preprocessing Routines. Yilai Li, Jennifer N.Cash, John J.G. Tesmer, Michael A.Cianfrocco. Structure 2020, Volume 28 (7), Pages 858-869.e3


CHANGES.txt
3.4: add discarded micrographs and use a generic updateOutputSet (by @danielmarchan3)
3.3: fix GPU assignment, add binary version detection
3.2:
    - protocol 2D assess now accepts as input 2D classes and outputs 2D classes
    - check if models exist
3.1.1 - add possible outputs class
3.1 - micassess updated to 1.0.0
3.0.4 - streaming support for micassess
3.0.3 - bugfixes
3.0.2 - add 0.2.0 version support
3.0.1 - fix README
3.0.0 - first version of the plugin


cryoassess/tests/__init__.py
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

from .test_protocols_cryoassess import TestCryoassess


cryoassess/tests/test_protocols_cryoassess.py
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

from pyworkflow.tests import BaseTest, DataSet, setupTestProject
from pyworkflow.utils import magentaStr
from pwem.protocols import ProtImportMicrographs, ProtImportAverages

from ..protocols import CryoassessProt2D, CryoassessProtMics


class TestCryoassess(BaseTest):
    @classmethod
    def setUpClass(cls):
        setupTestProject(cls)
        cls.ds = DataSet.getDataSet('relion_tutorial')
        cls.ds2 = DataSet.getDataSet('mda')
        cls.micsFn = cls.ds.getFile('micrographs/*mrc')
        cls.avgsFn = cls.ds2.getFile('averages/averages.stk')

        print(magentaStr("\n==> Importing data - micrographs:"))
        cls.protImportMics = cls.newProtocol(
            ProtImportMicrographs,
            filesPath=cls.micsFn,
            samplingRate=7.08)
        cls.launchProtocol(cls.protImportMics)

        print(magentaStr("\n==> Importing data - averages:"))
        cls.protImportAvgs = cls.newProtocol(
            ProtImportAverages,
            filesPath=cls.avgsFn,
            samplingRate=5.04,
            checkStack=True)
        cls.launchProtocol(cls.protImportAvgs)

    def testMicAssess(self):
        print(magentaStr("\n==> Testing cryoassess - micassess:"))
        protAssessMics = self.newProtocol(
            CryoassessProtMics,
            inputMicrographs=self.protImportMics.outputMicrographs)
        self.launchProtocol(protAssessMics)
        micSet = getattr(protAssessMics, 'outputMicrographs', None)
        self.assertEqual(len(micSet), self._getNumberOfGood(protAssessMics))

    def test2DAssess(self):
        print(magentaStr("\n==> Testing cryoassess - 2d assess:"))
        protAssess2D = self.newProtocol(
            CryoassessProt2D,
            inputRefs=self.protImportAvgs.outputAverages)
        self.launchProtocol(protAssess2D)
        avgSet = getattr(protAssess2D, 'outputAverages', None)
        self.assertIsNotNone(avgSet)

    def _getNumberOfGood(self, protMicAssess):
        with open(protMicAssess._getExtraPath('input_micrographs_good.star')) as f:
            lines = f.readlines()
        return len(lines) - 4


cryoassess/bibtex.py
# coding: utf-8
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************
"""

@article{Li2020,
title = "High-Throughput Cryo-EM Enabled by User-Free Preprocessing Routines",
journal = "Structure",
volume = "28",
number = "7",
pages = "858 - 869.e3",
year = "2020",
issn = "0969-2126",
doi = "https://dx.doi.org/10.1016/j.str.2020.03.008",
url = "https://www.sciencedirect.com/science/article/pii/S0969212620300800",
author = "Yilai Li and Jennifer N. Cash and John J.G. Tesmer and Michael A. Cianfrocco",
}

"""


cryoassess/constants.py
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************


def getCryoAssessEnvName(version):
    return "cryoassess-%s" % version


V1_0_0 = "1.0.0"
VERSIONS = [V1_0_0]
CRYOASSESS_DEFAULT_VER_NUM = V1_0_0

DEFAULT_ENV_NAME = getCryoAssessEnvName(CRYOASSESS_DEFAULT_VER_NUM)
DEFAULT_ACTIVATION_CMD = 'conda activate ' + DEFAULT_ENV_NAME
CRYOASSESS_ENV_ACTIVATION = 'CRYOASSESS_ENV_ACTIVATION'
CRYOASSESS_MODELS = 'CRYOASSESS_MODELS'

REF_CLASSES = 0
REF_AVERAGES = 1
DISCARDED = '_discarded'


cryoassess/protocols/protocol_2dassess.py
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

from glob import glob
import re

from pyworkflow.constants import PROD
import pyworkflow.protocol.params as params
from pyworkflow.protocol import ProtStreamingBase
from pyworkflow.protocol.constants import STEPS_PARALLEL
from pwem.protocols import ProtProcessParticles
from pwem.objects import SetOfAverages, SetOfClasses2D

from .. import Plugin
from ..constants import *


class CryoassessProt2D(ProtProcessParticles, ProtStreamingBase):
    """
    Protocol to assess 2D classes and 2D averages
    """
    _label = 'assess 2D classes'
    _devStatus = PROD
    _possibleOutputs = {
        'outputAverages': SetOfAverages,
        'outputAverages_discarded': SetOfAverages,
        'outputClasses': SetOfClasses2D,
        'outputClasses_discarded': SetOfClasses2D
    }
    outputDict = {}

    def __init__(self, **kwargs):
        ProtProcessParticles.__init__(self, **kwargs)
        self.stepsExecutionMode = STEPS_PARALLEL

    def _createFilenameTemplates(self):
        """ Centralize how files are called. """
        myDict = {'input_cls': self._getExtraPath('input_classes.mrcs')}
        self._pathTemplate = self._getExtraPath('2DAssess/*/particle_*.jpg')
        self._goodClassesRegex = re.compile(r'Good/particle_(\d*)\.jpg')
        self._regex = re.compile(r'particle_(\d*)\.jpg')
        self.goodList = []
        self.badList = []

        self._updateFilenamesDict(myDict)

    # --------------------------- DEFINE param functions ----------------------
    def _defineParams(self, form):
        form.addSection(label='Input')
        form.addParam('inputRefs', params.PointerParam,
                      pointerClass='SetOfClasses2D, SetOfAverages',
                      label="Input references", important=True)
        form.addParam('batchSize', params.IntParam, default=32,
                      expertLevel=params.LEVEL_ADVANCED,
                      label='Batch size',
                      help='Batch size used in prediction. Default is 32. '
                           'Increasing this number will result in faster '
                           'prediction. If memory error/warning appears, '
                           'you should lower this number.')

        form.addParallelSection(threads=3, mpi=1)

    # --------------------------- INSERT steps functions ----------------------
    def stepsGeneratorStep(self) -> None:
        """
        This step should be implemented by any streaming protocol.
        It should check its input and when ready conditions are met
        call the self._insertFunctionStep method.
        """
        self._createFilenameTemplates()
        convertStep = self._insertFunctionStep('convertInputStep', prerequisites=[])
        run2DAssesStep = self._insertFunctionStep('run2DAssessStep', prerequisites=convertStep)
        self._insertFunctionStep('createOutputStep', prerequisites=run2DAssesStep)

    # --------------------------- STEPS functions -----------------------------
    def convertInputStep(self):
        """ Create a mrcs stack as expected by cryoassess and
        store a list of objIds."""
        imgSet = self.inputRefs.get()
        imgSet.writeStack(self._getFileName('input_cls'))

        self.clsIds = list(self.inputRefs.get().getIdSet())

    def run2DAssessStep(self):
        """ Call cryoassess with the appropriate parameters. """
        params = ' '.join(self._getArgs())
        program = Plugin.getProgram('2dassess')
        self.runJob(program, params, env=Plugin.getEnviron())

    def createOutputStep(self):
        inputRefs = self.inputRefs.get()
        if self._getRefsType() == REF_CLASSES:
            outRefs = SetOfClasses2D.create(self._getPath())
            outBadRefs = SetOfClasses2D.create(self._getPath(),
                                               suffix=DISCARDED)
        else:
            outRefs = self._createSetOfAverages()
            outBadRefs = self._createSetOfAverages(suffix=DISCARDED)

        outRefs.copyInfo(inputRefs)
        outBadRefs.copyInfo(inputRefs)

        # Search through output files and find good and bad classes
        self._getReferences()
        if len(self.goodList):
            if self._getRefsType() == REF_CLASSES:
                outRefs.setObjLabel('good 2D classes')
                outRefs.appendFromClasses(inputRefs,
                                          filterClassFunc=self._addGoodClass)
                self.outputDict['outputClasses'] = outRefs
            else:
                outRefs.setObjLabel('good class averages')
                outRefs.copyItems(inputRefs,
                                  updateItemCallback=self._addGoodAvg)
                self.outputDict['outputAverages'] = outRefs

        if len(self.badList):
            if self._getRefsType() == REF_CLASSES:
                outBadRefs.setObjLabel('bad 2D classes')
                outBadRefs.appendFromClasses(inputRefs,
                                             filterClassFunc=self._addBadClass)
                self.outputDict['outputClasses_discarded'] = outBadRefs
            else:
                outBadRefs.setObjLabel('bad class averages')
                outBadRefs.copyItems(inputRefs,
                                     updateItemCallback=self._addBadAvg)
                self.outputDict['outputAverages_discarded'] = outBadRefs

        self._defineOutputs(**self.outputDict)
        if self._getRefsType() == REF_CLASSES:
            if len(self.goodList):
                self._defineSourceRelation(self.inputRefs, outRefs)
            if len(self.badList):
                self._defineSourceRelation(self.inputRefs, outBadRefs)

    # --------------------------- INFO functions ------------------------------
    def _summary(self):
        summary = []
        if not self.isFinished():
            summary.append("Output not ready.")
        else:
            summary.append("Sorted class references into good and bad ones.")

        return summary

    # --------------------------- UTILS functions -----------------------------
    def _getArgs(self):
        """ Return the list of args for the command. """
        args = ['-i %s ' % self._getFileName('input_cls'),
                '-m %s/2dassess_062119.h5' % Plugin.getVar(CRYOASSESS_MODELS),
                '-b %d' % self.batchSize.get()]

        return args

    def _getReferences(self):
        """ Return the list of good and bad classes files. """
        files = sorted(glob(self._pathTemplate))
        for i in files:
            s = self._goodClassesRegex.search(i)
            if s:
                self.goodList.append(int(s.group(1)))
            else:
                s = self._regex.search(i)
                self.badList.append(int(s.group(1)))

    def _addGoodAvg(self, item, row):
        """ Callback function to append only good items. """
        if self.clsIds.index(item.getObjId()) + 1 not in self.goodList:
            setattr(item, "_appendItem", False)

    def _addBadAvg(self, item, row):
        """ Callback function to append only bad items. """
        if self.clsIds.index(item.getObjId()) + 1 not in self.badList:
            setattr(item, "_appendItem", False)

    def _addGoodClass(self, item):
        """ Callback function to append only good classes. """
        return False if self.clsIds.index(item.getObjId()) + 1 not in self.goodList else True

    def _addBadClass(self, item):
        """ Callback function to append only bad classes. """
        return False if self.clsIds.index(item.getObjId()) + 1 not in self.badList else True

    def _getRefsType(self):
        imgSet = self.inputRefs.get()
        return REF_CLASSES if isinstance(imgSet, SetOfClasses2D) else REF_AVERAGES


cryoassess/__init__.py
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

import os
import pwem
import pyworkflow.utils as pwutils
from pyworkflow import Config

from .constants import *


__version__ = '3.4'
_references = ['Li2020']
_logo = "cryoassess_logo.png"


class Plugin(pwem.Plugin):
    _pathVars = [CRYOASSESS_MODELS]
    _url = "https://github.com/scipion-em/scipion-em-cryoassess"
    _supportedVersions = VERSIONS

    @classmethod
    def _defineVariables(cls):
        cls._defineVar(CRYOASSESS_ENV_ACTIVATION, DEFAULT_ACTIVATION_CMD)
        cls._defineEmVar(CRYOASSESS_MODELS, 'cryoassess-models')

    @classmethod
    def getCryoAssessEnvActivation(cls):
        """ Remove the scipion home and activate the conda environment. """
        activation = cls.getVar(CRYOASSESS_ENV_ACTIVATION)
        scipionHome = Config.SCIPION_HOME + os.path.sep

        return activation.replace(scipionHome, "", 1)

    @classmethod
    def getActiveVersion(cls, *args):
        """ Return the env name that is currently active. """
        envVar = cls.getVar(CRYOASSESS_ENV_ACTIVATION)
        return envVar.split()[-1].split("-")[-1]

    @classmethod
    def getEnviron(cls):
        """ Setup the environment variables needed to launch cryoassess. """
        environ = pwutils.Environ(os.environ)
        if 'PYTHONPATH' in environ:
            # this is required for python virtual env to work
            del environ['PYTHONPATH']
        return environ

    @classmethod
    def getDependencies(cls):
        """ Return a list of dependencies. Include conda if
        activation command was not found. """
        condaActivationCmd = cls.getCondaActivationCmd()
        neededProgs = []
        if not condaActivationCmd:
            neededProgs.append('conda')

        return neededProgs

    @classmethod
    def defineBinaries(cls, env):
        for ver in VERSIONS:
            cls.addCryoAssessPackage(env, ver,
                                     default=ver == CRYOASSESS_DEFAULT_VER_NUM)

    @classmethod
    def addCryoAssessPackage(cls, env, version, default=False):
        CRYOASSESS_INSTALLED = 'cryoassess_%s_installed' % version
        ENV_NAME = getCryoAssessEnvName(version)

        # try to get CONDA activation command
        installCmd = [cls.getCondaActivationCmd()]

        # Create the environment
        installCmd.append('conda create -y -n %s -c anaconda -c conda-forge '
                          'python=3.7 cudatoolkit=11.2 cudnn=8.1;' % ENV_NAME)

        # Activate the new environment
        installCmd.append('conda activate %s;' % ENV_NAME)

        # Install downloaded code
        url = "https://github.com/cianfrocco-lab/Automatic-cryoEM-preprocessing.git"
        installCmd.extend(['git clone %s cryoassess-master &&'
                           'cd cryoassess-master && git checkout master &&'
                           'cd .. && pip install -e cryoassess-master &&' % url])

        # Flag installation finished
        installCmd.append('touch %s' % CRYOASSESS_INSTALLED)

        cryoassess_commands = [(" ".join(installCmd), CRYOASSESS_INSTALLED)]

        envPath = os.environ.get('PATH', "")
        # keep path since conda likely in there
        installEnvVars = {'PATH': envPath} if envPath else None
        env.addPackage('cryoassess', version=version,
                       tar='void.tgz',
                       commands=cryoassess_commands,
                       neededProgs=cls.getDependencies(),
                       default=default,
                       vars=installEnvVars)

    @classmethod
    def getProgram(cls, program):
        """ Create cryoAssess command line. """
        fullProgram = '%s %s && %s' % (cls.getCondaActivationCmd(),
                                       cls.getCryoAssessEnvActivation(),
                                       program)
        return fullProgram
